crates/bevy_utils/src/anystack.rs

```diff
@@ -0,0 +1,218 @@
+struct AnyStackMeta {
+    offset: usize,
+    func: unsafe fn(value: *mut u8, user_data: *mut u8),
+}
+
+pub struct AnyStack {
+    bytes: Vec<u8>,
```

Comment:
There's a bug I just realized.
If we reallocate this byte buffer there's a chance that that data will become unaligned. 
I'll probably have to manually implement a Vec... :/

Reply:
Hmm, would replacing some of the reads with read_unaligned also work?

Reply:
Another approach:
```rust
#[repr(transparent, align(16))]
struct AlignedBytes([u8; 16]);
```
and check that the alignment of each item on push is less than 16

Reply:
> and check that the alignment of each item on push is less than 16

Sadly I don't think this is a viable solution since they're could be command types which have alignment greater that 16. And we should definitly support all alignments.

> Hmm, would replacing some of the reads with read_unaligned also work?

That was my other thought. I'll probably go down this path (assuming its not UB haha...) Since that would make the implementation far simpler.

Reply:
An alternative is to just allocate the items with alignment greater than the threshold that we hit. The number of types in the wild with alignment greater than 16 is almost certainly insignification, so a second vector containing `*mut u8`s constructed from leaking boxes of the elements wouldn't be bad at all (deallocating with https://doc.rust-lang.org/beta/std/alloc/fn.dealloc.html)

I think the cost of the extra cognitive load the unsafe version which supports arbitrary alignments wins us is not worthwhile.

Reply:
>  so a second vector containing *mut u8s constructed from leaking boxes of the elements wouldn't be bad at all

I don't think this is the right solution. (If I understand what you're saying correctly), you're suggesting that we need to create `Box` for each element pushed (with alignment over 16), leak it, and then call `dealloc` on the contents later?

I think that's a decent solution, however, I'd strongly prefer to just keep the elements packed as tightly as possible and reduce the number of total `allocs` required. The whole point of this PR is to reduce the number of total allocs needed for the `CommandQueue`.

And even by introducing this, we have the same amount of `unsafe` code as the current version stands.

Reply:
It may just be best to keep this data structure for internal use only. 
E.g. rename it from `AnyStack` -> `CommandQueue` and then make it a private structure inside `bevy_ecs` instead of a public structure in `bevy_util`.
This way we're not exposing extra `unsafe` to the bevy users.

Reply:
That seems like a good idea.

Reply:
(also just as an aside @DJMcNab I really appreciate the thorough review :heart:)